### PR TITLE
logging/anomaly: Support configuration filter types

### PR DESCRIPTION
**Pull request by jlucovsky**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [2958](https://redmine.openinfosecfoundation.org/issues/2958)

Continuation of #4100

This PR changes the anomaly configuration options which will cause existing suricata-verify tests to fail. Thus, this PR requires [Suricata-verify PR #116](https://github.com/OISF/suricata-verify/pull/116)

Updates since last PR:
- Reworked configuration settings controlling anomaly logging (see discussion in #4100)
- Log records can be individually filtered by `"type"`: `decode`, `stream`, `application-layer`.

### CODE REVIEW

File: doc/userguide/output/eve/eve-json-format.rst [hunk 1]
```diff
@@ -122,15 +122,15 @@ when anomaly logging is enabled.
 Fields
 ------
 
-* "type": Either "packet", "stream" or "applayer". In rare cases, type will be "unknown".
+* "type": Either "decode", "stream" or "application-layer". In rare cases, type will be "unknown".
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-format.rst]: why the rename from ```applayer``` to ```application-layer```? I like the short version much better.

File: doc/userguide/output/eve/eve-json-output.rst [hunk 2]
```diff
@@ -86,19 +86,41 @@ Anomalies are event records created when packets with unexpected or anomalous
...
-            # Enable dumping of packet header
-            # packethdr: no            # enable dumping of packet header
+	- anomaly:
+		# Anomaly log records describe unexpected conditions such as truncated packets, packets with invalid
+		# IP/UDP/TCP length values, and other events that render the packet invalid for further processing
+		# or describe unexpected behavior on an established stream. Networks which experience high
+		# occurrences of anomalies may experience packet processing degradation.
+		#
+		# Anomalies are reported for the following:
+		# 1. Decode: Values and conditions that are detected while decoding individual packets. This includes
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-output.rst]: can you use shorter lines here?

File: src/output-json-anomaly.c [hunk 3]
```diff
@@ -87,7 +92,15 @@ static int AnomalyDecodeEventJson(ThreadVars *tv, JsonAnomalyLogThread *aft,
     char timebuf[64];
     CreateIsoTimeString(&p->ts, timebuf, sizeof(timebuf));
 
+    uint16_t log_type = aft->json_output_ctx->flags;
+    bool log_stream = log_type & LOG_JSON_STREAM_TYPE;
+    bool log_decode = log_type & LOG_JSON_DECODE_TYPE;
     for (int i = 0; i < p->events.cnt; i++) {
+        uint8_t event_code = p->events.events[i];
+        bool is_decode = EVENT_IS_DECODER_PACKET_ERROR(event_code);
+        if (is_decode && !log_decode) continue;
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: style: put continue on a new line

File: doc/userguide/output/eve/eve-json-output.rst (hunk 2 above)
Comment (jlucovsky) [reply doc/userguide/output/eve/eve-json-output.rst]: Yes; will revise.

File: doc/userguide/output/eve/eve-json-format.rst (hunk 1 above)
Comment (jlucovsky) [reply doc/userguide/output/eve/eve-json-format.rst]: Will change to use `applayer`

**Comment by jlucovsky** [comment]
Continued in #4192

**Pull request closed without merging**